lib/navigation/preprocess_encoders.hpp

```diff
@@ -1,25 +1,63 @@
+#include "consts.hpp"
+
 #include <array>
 #include <cstdint>
+#include <cmath>
 
 #include "core/types.hpp"
+#include "core/logger.hpp"
+
 
 namespace hyped::navigation {
 class EncodersPreprocessor {
  public:
-  EncodersPreprocessor();
-
-  core::EncoderData processData(const core::EncoderData encoder_data);
+  EncodersPreprocessor(core::ILogger &logger);
+  std::optional<core::EncoderData> processData(const core::EncoderData encoder_data);
 
  private:
-  core::EncoderData detectOutliers(const core::EncoderData encoder_data);
 
-  void checkReliable(const core::EncoderData &encoder_data);
+  /**
+   * @brief tidies up the received data by detecting the outliers or data received from faulty 
+     sensors and replaces them with median value of the dataset.
+   *
+   * @param encoder_data array containing the values received from various encoder sensors
+   * @return consistent data with no outliers
+   */
+  std::optional<core::EncoderData> detectOutliers(const core::EncoderData encoder_data);
+  
+  SensorChecks checkReliable();
+
+
+  template<std::size_t N>
+  core::Float getSpecificQuartile(const std::array<std::uint32_t, N> &reliable_data, const core::Float quartile_percent){
+
+    const core::Float quartile_index = (num_reliable_encoders_ - 1)*quartile_percent;
+    const std::uint8_t quartile_high = static_cast<std::uint8_t>(std::ceil(quartile_index));
+    const std::uint8_t quartile_low = static_cast<std::uint8_t>(std::floor(quartile_index));
+    const core::Float quartile = (reliable_data.at(quartile_high) + reliable_data.at(quartile_low))/2.0;
+    return quartile;
+}
+
+ template<std::size_t N>
+ Quartile getQuartiles(std::array<std::uint32_t, N> &reliable_data){
+    std::sort(reliable_data.begin(), reliable_data.end());
```

Comment:
Look at the logs and try figure out why it doesn't build any more. Hint: Look at this line and then look at the included headers at the top of this file to see what you might have been missing...